components/camel-xj/src/main/java/org/apache/camel/component/xj/XmlSourceHandlerFactoryImpl.java

```diff
@@ -0,0 +1,150 @@
+import java.io.InputStream;
+
+public class XmlSourceHandlerFactoryImpl implements SourceHandlerFactory {
+
+    private XMLConverterHelper converter = new XMLConverterHelper();
+    private boolean isFailOnNullBody = true;
+
+    public void setFailOnNullBody(boolean failOnNullBody) {
+        isFailOnNullBody = failOnNullBody;
+    }
+
+    public boolean isFailOnNullBody() {
+        return isFailOnNullBody;
+    }
+
+    @Override
+    public Source getSource(Exchange exchange) throws Exception {
+        // only convert to input stream if really needed
+        if (isInputStreamNeeded(exchange)) {
+            InputStream is = exchange.getIn().getBody(InputStream.class);
+            return getSource(exchange, is);
+        } else {
+            Object body = exchange.getIn().getBody();
+            return getSource(exchange, body);
+        }
+    }
+
+    /**
+     * Checks whether we need an {@link InputStream} to access the message body.
+     * <p/>
+     * Depending on the content in the message body, we may not need to convert
+     * to {@link InputStream}.
+     *
+     * @param exchange the current exchange
+     * @return <tt>true</tt> to convert to {@link InputStream} beforehand converting to {@link Source} afterwards.
+     */
+    protected boolean isInputStreamNeeded(Exchange exchange) {
+        Object body = exchange.getIn().getBody();
+        if (body == null) {
+            return false;
+        }
+
+        if (body instanceof InputStream) {
+            return true;
+        } else if (body instanceof Source) {
+            return false;
+        } else if (body instanceof String) {
+            return false;
+        } else if (body instanceof byte[]) {
+            return false;
+        } else if (body instanceof Node) {
+            return false;
+        } else if (exchange.getContext().getTypeConverterRegistry().lookup(Source.class, body.getClass()) != null) {
+            //there is a direct and hopefully optimized converter to Source
+            return false;
+        }
+        // yes an input stream is needed
+        return true;
+    }
+
+    /**
+     * Converts the inbound body to a {@link Source}, if the body is <b>not</b> already a {@link Source}.
+     * <p/>
+     * This implementation will prefer to source in the following order:
+     * <ul>
+     *   <li>StAX - If StAX is allowed</li>
+     *   <li>SAX - SAX as 2nd choice</li>
+     *   <li>Stream - Stream as 3rd choice</li>
+     *   <li>DOM - DOM as 4th choice</li>
+     * </ul>
+     */
+    protected Source getSource(Exchange exchange, Object body) {
+        // body may already be a source
+        if (body instanceof Source) {
+            return (Source) body;
+        }
+        Source source = null;
+        if (body != null) {
+            // try StAX if enabled
+            source = exchange.getContext().getTypeConverter().tryConvertTo(StAXSource.class, exchange, body);
+
+            if (source == null) {
```

Comment:
Lines 114-125 and 386-397 are copy-pasted. Is it possible to refactor this to method?

Reply:
obsolete.